Optimization of predicates with toYear/toYYYYMM based on a general solution

ZhiguoZh opened the pull request:
As is suggested in issue https://github.com/ClickHouse/ClickHouse/issues/15257, the function preimage is a general
solution to the optimization problem with predicates containing the
date and datetime converters. This commit implements the idea by
adding the new methods, hasInformationAboutPreimage and getPreimage,
to IFunction/IFunctionBase, and having the specific convert functions
define their own preimage. Moreover, we added a new pass in the
TreeOptimizer and a new AST visitor for in-place rewriting the AST
with the converters' preimage.

Specifically, the optimization is applied to toYear and toYYYYMM.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
In the earlier PRs (#50062, #50307), we used to propose an optimization pattern which transforms the predicates with toYear/toYYYYMM into its equivalent but converter-free form. This transformation could bring significant performance impact to some workloads, such as SSB. However, as issue #50628 indicated, these two PRs would introduce some issues which may results in incomplete query results, and as a result, they were reverted by #50629.

After investigation, we've successfully identified the root cause: in the previous implementation, the preimage of the converters are often represented as a date range. For example, _**toYear(col) = 2023**_ would be transformed as **_col >= "2023-01-01" AND col <= "2023-12-31"_**. This is a viable transformation when **_col_** is a **Date** column, however,  if **_col_** is of **DateTime**, this comparison would lead to the implicit data type conversions, which round the date literals to the start of the days, **"2023-01-01 00:00:00"** and **"2023-12-31 00:00:00"**. It can be predicated, when we evaluate the predicate, **col <= "2023-12-31"**, the results on the last day would be filtered out.

To fix this issue, we majorly made two changes: 1) we define the preimage with a left-closed and right-open interval, instead of a bounded one, 2) we generate the string literals containing **Date** or **DateTime** according to the data type of converters' arguments. And the correctness could be validated with the functional tests (the originals and the new ones) in [date_predicate_optimizations](https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02783_date_predicate_optimizations.sql).

Moreover, inspired by #15257, we tried to implemented a framework of the general solution to this kind of problems with the per-function preimage. We added new interfaces, **hasInformationAboutPreimage** and **getPreimage**, to **IFunction/IFunctionBase**, and had the specific convert functions define their own preimages. In this way, the AST visitor would not bother handling the converter-specific details.

Currently, we intended to shrink the impacted functions to **FunctionDateOrDateTimeToSomething**. We plan to extend this design to others if this change is accepted.

On the ICX server (Intel Xeon Platinum 8380 CPU, 80 cores, 160 threads), this optimization could improve the **overall QPS** of **SSB** workloads by **10.6%** and some of them exceeding 30%.

![image](https://github.com/ClickHouse/ClickHouse/assets/26588299/2974c529-8329-44db-8553-c4b6ee9fad9e)


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

ZhiguoZh committed:
Add tests in date_predicate_optimizations

This commit adds tests of toYear/toYYYYMM for all of the date
types, including Date, Date32, DateTime and DateTime64, to ensure
the optimization for the date or datetime filter would not break
the queries. The test data are especially selected around the margins
of a year or a month.

ZhiguoZh committed:
The general optimization of predicates with date/datetime converters

As is suggested in issue #15257, the function preimage is a general
solution to the optimization problem with predicates containing the
date and datetime converters. This commit implements the idea by
adding the new methods, hasInformationAboutPreimage and getPreimage,
to IFunction/IFunctionBase, and having the specific convert functions
define their own preimage. Moreover, we added a new pass in the
TreeOptimizer and a new AST visitor for in-place rewriting the AST
with the converters' preimage.

Specifically, the optimization is applied to toYear and toYYYYMM.

hanfei1991 committed:
fix style

UnamedRus commented:
That's really great, thank you for your work

ZhiguoZh commented:
Thanks @UnamedRus! And I'd appreciate you for linking me to your discussion with Alexey, which really inspired me!

I hope this solution could soon extend to the problems you identified in https://github.com/ClickHouse/ClickHouse/issues/15257.

ZhiguoZh commented:
Hi @hanfei1991, it seems the 2 failed tests are unrelated to this change. This PR is ready for your review. Thanks!

hanfei1991 commented on src/Functions/FunctionDateOrDateTimeToSomething.h:
```diff
@@ -83,6 +83,13 @@ class FunctionDateOrDateTimeToSomething : public IFunctionDateOrDateTime<Transfo
+
+    RangeOrNull getPreimage(const IDataType & type, const Field & point) const override
+    {
+        return Transform::getPreimage(type, point);
```
we can define this function like
```
if constexpr (Transform::hasPregimage()) {return getPreimage ...}
else return std::nullopt;
```
so that we don't have to declear so many `getPreimage`

Of course, use `static constexpr bool hasPreimage` for each Transform class

ZhiguoZh replied:
Exactly! This constexpr-if statement would save a lot of code size. Thanks for your suggestion!

My only concern is the return of `std::nullopt` would lead to the silent errors when `FunctionDateOrDateTimeToSomething::hasInformationAboutPreimage() == false` and  `FunctionDateOrDateTimeToSomething::getPreimage()` is called. Maybe we should throw an exception here as we did for other similar cases?

Algunenano commented:
Will this work with the new analyzer too (`allow_experimental_analyzer=1`) or does it need to be implemented there too?

ZhiguoZh committed:
fix style

Move error throws in Transform to FunctionDateOrDateTimeToSomething.

ZhiguoZh commented:
> Will this work with the new analyzer too (`allow_experimental_analyzer=1`) or does it need to be implemented there too?

Hi @Algunenano, I've confirmed that this change does not work on the new analyzer, and I think it is needed to implement a new pass in the analyzer for this optimization. I will start the work in another PR once this change is accepted. Thanks for your advice!

hanfei1991 commented on src/Interpreters/OptimizeDateOrDateTimeConverterWithPreimageVisitor.cpp:
```diff
@@ -0,0 +1,188 @@
+                    std::make_shared<ASTLiteral>(end_date_or_date_time)
+                    );
+    }
+    else // comparator == "less" || comparator == "greaterOrEquals"
```
it's more safe to throw LOGICAL_ERROR exception when it's not less or greater or equal.
but it's up to you, just suggestion

ZhiguoZh committed:
Enhance safety of function generateOptimizedDateFilterAST

This commit checks the corner case where the comparator is none
of equals, notEquals, less, lessOrEquals, greater, greaterOrEquals,
and throws LOGICAL_ERROR exception if so.

ZhiguoZh commented:
Thanks so much for your code review and acceptance of this PR!

alexey-milovidov commented:
The test appears to behave differently depending on the time zone. It makes me suspect that the optimization is incorrect.
A fix (or resubmit) is needed. See #51390

ZhiguoZh commented:
> The test appears to behave differently depending on the time zone. It makes me suspect that the optimization is incorrect. A fix (or resubmit) is needed. See #51390

Thanks Alexey! I will look into this issue and try to find a fix. @alexey-milovidov
